code/modules/mob/living/carbon/xenomorph/castes/warrior/abilities_warrior.dm

```diff
@@ -520,3 +514,158 @@
+
+/datum/action/xeno_action/activable/punch/weak
+	mechanics_text = "Strike a target, weaker than a regular punch. Resets jab cooldown."
+	plasma_cost = 10
+
+/datum/action/xeno_action/activable/punch/weak/use_ability(atom/A)
+	var/mob/living/carbon/xenomorph/X = owner
+	var/damage = X.xeno_caste.melee_damage * X.xeno_melee_damage_modifier * 0.5
+	var/target_zone = check_zone(X.zone_selected)
+
+	if(isliving(A))
+		var/mob/living/target = A
+		if(!target.punch_act(X, damage, target_zone, push = FALSE,  punch_description = "swift")) //Doesn't push away so it's easier to combo with jab
+			return fail_activate()
+	else
+		if(!A.punch_act(X, damage, target_zone))
+			return fail_activate()
+
+	GLOB.round_statistics.warrior_punches++
+	SSblackbox.record_feedback("tally", "round_statistics", 1, "warrior_punches")
+
+	if(isliving(A))
+		var/mob/living/target = A
+		var/datum/action/xeno_action/jab = X.actions_by_path[/datum/action/xeno_action/activable/punch/jab]
+		jab.clear_cooldown()
+
+		add_to_combo(target)
+
+	succeed_activate()
+	add_cooldown()
+
+// ***************************************
+// *********** Uppercut
+// ***************************************
+
+/datum/action/xeno_action/activable/uppercut
+	name = "Uppercut"
+	action_icon_state = "uppercut"
+	mechanics_text = "Inflicts debilitating damage to a nearby target depending on how many times you recently hit them with jab and punch."
+	ability_name = "uppercut"
+	plasma_cost = 12
+	cooldown_timer = 10 SECONDS
+	target_flags = XABB_MOB_TARGET
+	keybind_signal = COMSIG_XENOABILITY_UPPERCUT
+	var/combo[0] //Who did we punch/jab recently and how many times ?
+
+/datum/action/xeno_action/activable/uppercut/can_use_ability(atom/A, silent = FALSE, override_flags)
+	.=..()
+
+	var/mob/living/L = A
+	if(!isliving(A) || isxeno(A) || L.stat == DEAD)
+		if(!silent)
+			to_chat(owner, span_xenodanger("We cannot uppercut this target!"))
+		return FALSE
+
+	if(!owner.Adjacent(A))
+		if(!silent)
+			to_chat(owner, span_xenodanger("Our target must be closer!"))
+
+/datum/action/xeno_action/activable/uppercut/use_ability(atom/A)
+	var/mob/living/carbon/xenomorph/X = owner
+	var/mob/living/target = A
+	var/current_combo = combo[target.name]
+	var/damage = min(X.xeno_caste.melee_damage +  X.xeno_caste.melee_damage * (current_combo/2), 200)
+	var/cameraShake = min(1/2 * current_combo, 4)
+	X.face_atom(A)
+	X.do_attack_animation(A, ATTACK_EFFECT_SMASH)
+	X.do_attack_animation(A, ATTACK_EFFECT_PUNCH)
+
+	X.visible_message(span_xenodanger("\The [X] uppercuts [A]!"), \
+		span_xenodanger("We uppercut [A] with all our might!"), visible_message_flags = COMBAT_MESSAGE)
+
+	if(current_combo >= WARRIOR_WEAK_COMBO_THRESHOLD)
+		target.KnockdownNoChain(1 SECONDS)
+	if(current_combo >= WARRIOR_AVERAGE_COMBO_THRESHOLD)
+		target.ParalyzeNoChain(2 SECONDS)
+	if(current_combo >= WARRIOR_PERFECT_COMBO_THRESHOLD)
+		target.Sleeping(4 SECONDS, ignore_canstun = TRUE)
+		//Figure out a good animation
```

Comment:
todo list?

Also as a side note I'd like to discuss these stuns @ me in discord